### ISSUE TITLE
Question: How to best set up model when we expect missing data for a given feature

### ISSUE THREAD

**Issue by Fish-Soup** [microsoft/LightGBM#4477]
### Context
I am working on some time series forecasting, where normally I can use exogeneous variables to forecast well.  However often the addition of lagged target variables to the features is helpful. 

Short term forecasting is the important but long term forecasting is as well. I could fit 2 models, a short term one with a target-1 feature added and long term without. However this seems wasteful as most of the features are the same. I know lighGBM can handle missing data so I wanted to experiment with this.  To do a little test I took the boston data from sklearn and made it have a lagged component in the target.

```python
data = load_boston()
y = pd.Series(data["target"])
X = pd.DataFrame(data=data["data"], columns=data["feature_names"])

fac = 0.5
for i in range(1, len(y)):
    y[i] = (1 - fac) * y[i] + y[i - 1] * fac
```

### Basic Idea
My idea was to double the length of my dataset, with half the data points getting this extra target-1 feature and the other half not.  My understanding is lightGBM treats missing variables differently if they have been missing in training or not. To make sure I was understanding what was happening I thought I decided to fit 3 models to different samples of the training data
1. Only data without target-1 feature (sample_set==long)
2. Only data with the target-1 feature (sample_set ==1)
3. Data with and without the target-1 feature. (sample_set== long+1)
On these 3 models I predicted test data with out the target-1 (lag=long) and with the target-1 (lag=1) feature.

### Results

| sample_set| lag| mse|
| ------------- |:-------------:| -----:|
| long      | 1 | 35.174427 |
| long     | long      |   35.174427 |
| 1 | 1      | 11.624387 |
| 1 | long      |    28.090413|
| long+1 | 1      |   24.113144 |
| long+1 | long      |    36.401559 |

So its not surprising 
1. when sample_set = long that the error on lags 1 and long is the same, 
2. when sample_set= long+1 and lag=long that the error is about that same as when sample_set=long
3. when sample_set= long+1 and lag=1 has a lower error than lag=long
4. when sample_set= 1 and lag=1 has a lower error than lag=long

However the 2 things I found surprising.
1. sample_set=long+1, lag=1 has a significantly higher error than sample_set=1, lag=1
2. sample_set=long+1, lag=long has a significantly higher error than sample_set=1, lag=long.

I especially don't understand the 2nd point as it seems lightGBM is handling the missing data better when it hasn't been exposed to missing data before.

### Final Thoughts
I haven't done any hyper-parameter tuning so this may be driving these strange scores.

**Comment by shiyu1994** [comment]
@Fish-Soup Thanks for using LightGBM.
Does the lag=long and lag=1 have very different label values? Or does the two tasks have very different ground truth distribution? If so, combining the training set the two may not be helpful if the goal is to predict for only lag=long or lag=1.

Though LightGBM can handle missing values, that does not mean that it can handle two very different learning tasks in the same  time.

**Comment by shiyu1994** [comment]
Please feel free to post here if you have any further questions or comments.

**Comment by Fish-Soup** [comment]
@shiyu1994 thanks for responding. Not sure if I made it clear but this is a regression problem. 

The only difference between the lag=long and the lag=1 data is that the later has values in the target-1 feature whilst the former has nans. 

This means when we fit on sample_set=1 we fit only on the data thats has the target-1 feature. In this case lightGBM gets the best accuracy both the case with target-1 has values and is missing. 

I had presumed  that lightGBM would do better predicting lag=long (when target-1 feature is nan), if it had been fit on data missing this feature. 
But this does not seem to be the case.

Many thanks.
